File: erlang/space-age/space_age.erl
-module(space_age).
-export([ageOn/2]).

ageOn(earth,   Seconds) -> age_ratio(Seconds, 1);
ageOn(mercury, Seconds) -> age_ratio(Seconds, 0.2408467);
ageOn(venus,   Seconds) -> age_ratio(Seconds, 0.61519726);
ageOn(mars,    Seconds) -> age_ratio(Seconds, 1.8808158);
ageOn(jupiter, Seconds) -> age_ratio(Seconds, 11.862615);
ageOn(saturn,  Seconds) -> age_ratio(Seconds, 29.447498);
ageOn(uranus,  Seconds) -> age_ratio(Seconds, 84.016846);
ageOn(neptune, Seconds) -> age_ratio(Seconds, 164.79132);
ageOn(_, _) -> error.

age_ratio(Seconds, Ratio) ->
    (Seconds / (31557600 * Ratio)).

File: haskell/space-age/src/SpaceAge.hs
module SpaceAge (Planet(..), ageOn) where

data Planet = Mercury
            | Venus
            | Earth
            | Mars
            | Jupiter
            | Saturn
            | Uranus
            | Neptune

earthSeconds = 31557600

ageOn :: Planet -> Float -> Float
ageOn planet seconds =
  case planet of
    Mercury -> relativeAge 0.2408467 seconds
    Venus -> relativeAge 0.61519726 seconds
    Earth -> relativeAge 1 seconds
    Mars -> relativeAge 1.8808158 seconds
    Jupiter -> relativeAge 11.862615 seconds
    Saturn -> relativeAge 29.447598 seconds
    Uranus -> relativeAge 84.016846 seconds
    Neptune -> relativeAge 164.79132 seconds

relativeAge :: Float -> Float -> Float
relativeAge ratio seconds =
    (seconds / earthSeconds) / ratio
